<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.53">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Roy Francis">
<meta name="description" content="Comparison of different integration methods. Integration of 6 human breast celltypes from 7 donors">

<title>Breast</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="breast_files/libs/clipboard/clipboard.min.js"></script>
<script src="breast_files/libs/quarto-html/quarto.js"></script>
<script src="breast_files/libs/quarto-html/popper.min.js"></script>
<script src="breast_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="breast_files/libs/quarto-html/anchor.min.js"></script>
<link href="breast_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="breast_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="breast_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="breast_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="breast_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="breast_files/libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="breast_files/libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="breast_files/libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">


</head>

<body>

<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-left">
      <h1 class="title">Breast</h1>
            <p class="subtitle lead">Single-Cell Integration</p>
                  <div>
        <div class="description">
          Comparison of different integration methods. Integration of 6 human breast celltypes from 7 donors
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-left">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Roy Francis </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">22-09-2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-full">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#original" id="toc-original" class="nav-link active" data-scroll-target="#original"><span class="header-section-number">1</span> Original</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods"><span class="header-section-number">2</span> Methods</a>
  <ul class="collapse">
  <li><a href="#seurat-cca" id="toc-seurat-cca" class="nav-link" data-scroll-target="#seurat-cca"><span class="header-section-number">2.1</span> Seurat CCA</a></li>
  <li><a href="#seurat-cca-sct" id="toc-seurat-cca-sct" class="nav-link" data-scroll-target="#seurat-cca-sct"><span class="header-section-number">2.2</span> Seurat CCA SCT</a></li>
  <li><a href="#seurat-rpca" id="toc-seurat-rpca" class="nav-link" data-scroll-target="#seurat-rpca"><span class="header-section-number">2.3</span> Seurat RPCA</a></li>
  <li><a href="#harmony" id="toc-harmony" class="nav-link" data-scroll-target="#harmony"><span class="header-section-number">2.4</span> Harmony</a></li>
  <li><a href="#liger" id="toc-liger" class="nav-link" data-scroll-target="#liger"><span class="header-section-number">2.5</span> Liger</a></li>
  <li><a href="#fastmnn" id="toc-fastmnn" class="nav-link" data-scroll-target="#fastmnn"><span class="header-section-number">2.6</span> FastMNN</a></li>
  <li><a href="#stacas" id="toc-stacas" class="nav-link" data-scroll-target="#stacas"><span class="header-section-number">2.7</span> STACAS</a></li>
  <li><a href="#stacas-grp" id="toc-stacas-grp" class="nav-link" data-scroll-target="#stacas-grp"><span class="header-section-number">2.8</span> STACAS Grp</a></li>
  <li><a href="#conos" id="toc-conos" class="nav-link" data-scroll-target="#conos"><span class="header-section-number">2.9</span> Conos</a></li>
  <li><a href="#scanorama" id="toc-scanorama" class="nav-link" data-scroll-target="#scanorama"><span class="header-section-number">2.10</span> Scanorama</a></li>
  <li><a href="#scvi" id="toc-scvi" class="nav-link" data-scroll-target="#scvi"><span class="header-section-number">2.11</span> scVI</a></li>
  </ul></li>
  <li><a href="#compare" id="toc-compare" class="nav-link" data-scroll-target="#compare"><span class="header-section-number">3</span> Compare</a>
  <ul class="collapse">
  <li><a href="#umap" id="toc-umap" class="nav-link" data-scroll-target="#umap"><span class="header-section-number">3.1</span> UMAP</a></li>
  <li><a href="#metrics" id="toc-metrics" class="nav-link" data-scroll-target="#metrics"><span class="header-section-number">3.2</span> Metrics</a></li>
  <li><a href="#time-memory" id="toc-time-memory" class="nav-link" data-scroll-target="#time-memory"><span class="header-section-number">3.3</span> Time &amp; Memory</a></li>
  </ul></li>
  <li><a href="#versions" id="toc-versions" class="nav-link" data-scroll-target="#versions"><span class="header-section-number">4</span> Versions</a></li>
  <li><a href="#session" id="toc-session" class="nav-link" data-scroll-target="#session"><span class="header-section-number">5</span> Session</a></li>
  </ul>
</nav>
</div>
<main class="content quarto-banner-title-block column-page-left" id="quarto-document-content">





<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(dplyr)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># library(tidyr)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># library(ggpubr)</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(randomcoloR)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("tpq/peakRAM")</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(peakRAM)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github('satijalab/seurat-data')</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># library(SeuratData)</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># InstallData("panc8")</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github('satijalab/seurat-wrappers')</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratWrappers)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("hms-dbmi/conos")</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># library(conos)</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("immunogenomics/harmony")</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># library(harmony)</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("welch-lab/liger")</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co"># library(rliger)</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("carmonalab/STACAS")</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="co"># library(STACAS)</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("cellgeni/sceasy")</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="co"># library(sceasy)</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="co"># library(batchelor)</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="co"># scanorama, scvi-tools</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co"># library(reticulate)</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("carmonalab/scIntegrationMetrics")</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scIntegrationMetrics)</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plummy)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(showtext)</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>plummy<span class="sc">::</span><span class="fu">import_barlow</span>()</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>showtext<span class="sc">::</span><span class="fu">showtext_auto</span>()</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>showtext<span class="sc">::</span><span class="fu">showtext_opts</span>(<span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a><span class="co"># set future globals size to 6.2GB</span></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="fl">6.2</span> <span class="sc">*</span> <span class="dv">1024</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>suffix <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"original"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"cca"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"cca-sct"</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"rpca"</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"fastmnn"</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"harmony"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"liger"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"conos"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"stacas"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"stacas-grp"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"scanorama"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"scvi"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>method_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Original"</span>, <span class="st">"Seurat CCA"</span>, <span class="st">"Seurat CCA SCT"</span>, <span class="st">"Seurat RPCA"</span>, <span class="st">"FastMNN"</span>, <span class="st">"Harmony"</span>, <span class="st">"Liger"</span>, <span class="st">"Conos"</span>, <span class="st">"STACAS"</span>, <span class="st">"STACAS Grp"</span>, <span class="st">"Scanorama"</span>, <span class="st">"scVI"</span>)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>path_rds <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-"</span>, suffix, <span class="st">".rds"</span>))</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>path_memtime <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-"</span>, suffix, <span class="st">".rds"</span>))</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>path_metrics <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-"</span>, suffix, <span class="st">".rds"</span>))</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>plt <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, <span class="at">red =</span> <span class="st">"umap"</span>, <span class="at">grp =</span> <span class="st">"batch"</span>, <span class="at">label =</span> <span class="cn">NULL</span>, <span class="at">cols =</span> <span class="cn">NULL</span>) {</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># cols &lt;- c("#f1948a", "#c39bd3", "#85c1e9", "#a2d9ce", "#f9e79f", "#fad7a0", "#e74c3c", "#8e44ad", "#3498db", "#138d75", "#27ae60", "#f1c40f", "#f39c12", "#e67e22", "#95a5a6", "#5d6d7e")</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(obj[[]][, grp]))</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(cols)) cols <span class="ot">&lt;-</span> randomcoloR<span class="sc">::</span><span class="fu">distinctColorPalette</span>(l)</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  lpos <span class="ot">&lt;-</span> <span class="st">"top"</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (l <span class="sc">&gt;</span> <span class="dv">30</span>) lpos <span class="ot">&lt;-</span> <span class="st">"none"</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(obj, <span class="at">group.by =</span> grp, <span class="at">reduction =</span> red, <span class="at">pt.size =</span> <span class="fl">0.3</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> cols) <span class="sc">+</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> label) <span class="sc">+</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_umap</span>() <span class="sc">+</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> lpos,</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj)</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x)</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>fn_plots <span class="ot">&lt;-</span> <span class="cf">function</span>(path, <span class="at">red =</span> <span class="st">"umap"</span>, <span class="at">grp =</span> <span class="st">"celltype"</span>, <span class="at">label =</span> <span class="cn">NULL</span>, <span class="at">path_umap =</span> <span class="cn">NULL</span>, <span class="at">cols =</span> <span class="cn">NULL</span>, <span class="at">export =</span> <span class="cn">FALSE</span>) {</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>  showtext<span class="sc">::</span><span class="fu">showtext_auto</span>()</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>  showtext<span class="sc">::</span><span class="fu">showtext_opts</span>(<span class="at">dpi =</span> <span class="dv">300</span>)</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(label)) <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"Running "</span>, label, <span class="st">"..."</span>))</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">plt</span>(<span class="fu">readRDS</span>(path), <span class="at">red =</span> red, <span class="at">grp =</span> grp, <span class="at">label =</span> label, <span class="at">cols =</span> cols)</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (export) {</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"Exporting "</span>, path_umap))</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggsave</span>(path_umap, x, <span class="at">height =</span> <span class="fl">5.5</span>, <span class="at">width =</span> <span class="dv">5</span>)</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x)</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>fn_memtime <span class="ot">&lt;-</span> <span class="cf">function</span>(path, <span class="at">label =</span> <span class="cn">NULL</span>) {</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>  dfr <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(path)[, <span class="fu">c</span>(<span class="st">"Elapsed_Time_sec"</span>, <span class="st">"Total_RAM_Used_MiB"</span>, <span class="st">"Peak_RAM_Used_MiB"</span>)]</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(dfr) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"time_sec"</span>, <span class="st">"total_ram_mb"</span>, <span class="st">"peak_ram_mb"</span>)</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>  dfr<span class="sc">$</span>method <span class="ot">&lt;-</span> label</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(dfr)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>fn_metrics <span class="ot">&lt;-</span> <span class="cf">function</span>(path, <span class="at">label =</span> <span class="cn">NULL</span>) {</span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>  dfr <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">readRDS</span>(path))</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>  dfr<span class="sc">$</span>method <span class="ot">&lt;-</span> label</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(dfr)</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a><span class="co">#' @description Get integration metrics</span></span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param obj Seurat v5 object</span></span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param batch Batch variable</span></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param grp Group variable (celltype)</span></span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param metrics_ilp Perplexity value</span></span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param metrics_red Reduction to use</span></span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param n Max number of cells per cluster</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a><span class="co">#'</span></span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>fn_getmetrics <span class="ot">&lt;-</span> <span class="cf">function</span>(obj, batch, grp, <span class="at">metrics_ilp=</span><span class="dv">21</span>, <span class="at">metrics_red=</span><span class="st">"umap"</span>,<span class="at">n=</span><span class="dv">5000</span>) {</span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>  tbl <span class="ot">&lt;-</span> <span class="fu">table</span>(obj[[]][, <span class="fu">c</span>(grp, batch)])</span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">any</span>(tbl <span class="sc">==</span> <span class="dv">0</span>)) {</span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Removing group-batches with zero counts...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>    obj <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, <span class="at">subset =</span> <span class="sc">!!</span><span class="fu">sym</span>(grp) <span class="sc">%in%</span> <span class="fu">names</span>(<span class="fu">which</span>(<span class="fu">rowSums</span>(tbl <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> <span class="fu">ncol</span>(tbl))))</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Idents</span>(obj) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(obj[[]][,batch])</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">any</span>(<span class="fu">table</span>(obj[[]][,batch])<span class="sc">&gt;</span>n)) {</span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Number of cells per "</span>,batch,<span class="st">" larger than "</span>,n,<span class="st">". Downsampling...</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>    obj <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, <span class="at">downsample =</span> n)</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>  obj<span class="sc">@</span>meta.data[] <span class="ot">&lt;-</span> <span class="fu">lapply</span>(obj<span class="sc">@</span>meta.data, <span class="cf">function</span>(x) <span class="cf">if</span> (<span class="fu">is.factor</span>(x)) <span class="fu">droplevels</span>(x) <span class="cf">else</span> x)</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Preview:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(obj)</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Counts by group and batch:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">table</span>(obj[[]][, <span class="fu">c</span>(grp, batch)]))</span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">getIntegrationMetrics</span>(obj,</span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta.label =</span> grp,</span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta.batch =</span> batch,</span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>    <span class="at">iLISI_perplexity =</span> metrics_ilp,</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a>    <span class="at">method.reduction =</span> metrics_red</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a>theme_umap <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_family =</span> <span class="st">"barlow"</span>, <span class="at">base_size =</span> <span class="dv">9</span>) <span class="sc">+</span></span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb2-108"><a href="#cb2-108" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-109"><a href="#cb2-109" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-110"><a href="#cb2-110" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-111"><a href="#cb2-111" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-112"><a href="#cb2-112" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">10</span>, <span class="dv">10</span>),</span>
<span id="cb2-113"><a href="#cb2-113" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb2-114"><a href="#cb2-114" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb2-115"><a href="#cb2-115" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">color =</span> <span class="st">"white"</span>),</span>
<span id="cb2-116"><a href="#cb2-116" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.key.spacing.y =</span> <span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">"cm"</span>)</span>
<span id="cb2-117"><a href="#cb2-117" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-118"><a href="#cb2-118" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-119"><a href="#cb2-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-120"><a href="#cb2-120" aria-hidden="true" tabindex="-1"></a>theme_report <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb2-121"><a href="#cb2-121" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_family =</span> <span class="st">"barlow"</span>, <span class="at">base_size =</span> <span class="dv">9</span>) <span class="sc">+</span></span>
<span id="cb2-122"><a href="#cb2-122" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb2-123"><a href="#cb2-123" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="dv">15</span>, <span class="dv">15</span>, <span class="dv">15</span>, <span class="dv">15</span>),</span>
<span id="cb2-124"><a href="#cb2-124" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">color =</span> <span class="st">"white"</span>)</span>
<span id="cb2-125"><a href="#cb2-125" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-126"><a href="#cb2-126" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-127"><a href="#cb2-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-128"><a href="#cb2-128" aria-hidden="true" tabindex="-1"></a>cols_bright <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#FF363CFF"</span>, <span class="st">"#6EE2FFFF"</span>, <span class="st">"#F7C530FF"</span>, <span class="st">"#95CC5EFF"</span>, <span class="st">"#D0DFE6FF"</span>, <span class="st">"#F79D1EFF"</span>, <span class="st">"#748AA6FF"</span>, <span class="st">"#017A4AFF"</span>, <span class="st">"#3D98D3FF"</span>, <span class="st">"#7559A2FF"</span>, <span class="st">"#794924FF"</span>, <span class="st">"#A6CEE3"</span>, <span class="st">"#34495e"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Parameters defined in the report:</p>
<p><strong>label</strong>: breast <br><strong>batch</strong>: donor_id <br><strong>grp</strong>: cell_type <br><strong>dims_umap_max</strong>: 20 <br><strong>metrics_ilp</strong>: 21 <br><strong>metrics_red</strong>: umap <br><strong>path</strong>: . <br></p>
<section id="original" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="original"><span class="header-section-number">1</span> Original</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path,<span class="st">"obj.rds"</span>))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># copy the batch variable as batch</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>batch <span class="ot">&lt;-</span> obj[[]][,batch]</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
22593 features across 31696 samples within 1 assay 
Active assay: RNA (22593 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(obj[[]][,batch])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
                      D1                       D2                       D3 
                    2303                      864                     2517 
                      D4                       D5                      D11 
                    1771                     2244                     7454 
pooled [D9,D7,D8,D10,D6] 
                   14543 </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove batches less than 30 cells</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dfr <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">table</span>(obj[[]][,batch]))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>dfr1 <span class="ot">&lt;-</span> dfr <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(Freq<span class="sc">&gt;</span><span class="dv">30</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">nrow</span>(dfr)<span class="sc">!=</span><span class="fu">nrow</span>(dfr1)) {</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"Of "</span>,<span class="fu">nrow</span>(dfr),<span class="st">" batches, "</span>,<span class="fu">paste0</span>(<span class="fu">nrow</span>(dfr)<span class="sc">-</span><span class="fu">nrow</span>(dfr1),<span class="st">" batches discarded because they contain less than 30 cells. Final number of batches: "</span>,<span class="fu">nrow</span>(dfr1),<span class="st">"."</span>))</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj,<span class="at">subset=</span>batch <span class="sc">%in%</span> dfr1<span class="sc">$</span>Var1)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  obj</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">table</span>(obj[[]][,batch])</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co"># create and save color palettes the first time</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path,<span class="fu">paste0</span>(label,<span class="st">"-seurat-original.rds"</span>)))){</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># if grp length is more than 13, create distinct colors</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">unique</span>(obj[[]][,grp])) <span class="sc">&gt;</span> <span class="fu">length</span>(cols_bright)){</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    cols_grp <span class="ot">&lt;-</span> randomcoloR<span class="sc">::</span><span class="fu">distinctColorPalette</span>(<span class="fu">length</span>(<span class="fu">unique</span>(obj[[]][,grp])))</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  }<span class="cf">else</span>(</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    cols_grp <span class="ot">&lt;-</span> cols_bright</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>  cols_batch <span class="ot">&lt;-</span> randomcoloR<span class="sc">::</span><span class="fu">distinctColorPalette</span>(<span class="fu">length</span>(<span class="fu">unique</span>(obj[[]][,batch])))</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(cols_grp,<span class="fu">file.path</span>(path,<span class="st">"colors-grp.rds"</span>))</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(cols_batch,<span class="fu">file.path</span>(path,<span class="st">"colors-batch.rds"</span>))</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span>{</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>  cols_grp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path,<span class="st">"colors-grp.rds"</span>))</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>  cols_batch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path,<span class="st">"colors-batch.rds"</span>))</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(dfr,dfr1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-original.rds"</span>)))) {</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">FindVariableFeatures</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunPCA</span>(<span class="at">npcs =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1,<span class="at">batch=</span>batch,<span class="at">grp=</span>grp,<span class="at">metrics_ilp=</span>metrics_ilp,<span class="at">metrics_red=</span>metrics_red)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-original.rds"</span>)))</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-original.rds"</span>)))</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-original.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-original.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-original.rds"</span>)))</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1   D2   D3
  endothelial cell of lymphatic vessel                         26   15   11
  basal-myoepithelial cell of mammary gland                   232   60  583
  luminal adaptive secretory precursor cell of mammary gland 1264  460 1109
  luminal hormone-sensing cell of mammary gland               777  329  809
                                                            donor_id
cell_type                                                      D4   D5  D11
  endothelial cell of lymphatic vessel                         10    4   19
  basal-myoepithelial cell of mammary gland                   174  326 2305
  luminal adaptive secretory precursor cell of mammary gland  842 1193 2560
  luminal hormone-sensing cell of mammary gland               744  719  116
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  endothelial cell of lymphatic vessel                                             15
  basal-myoepithelial cell of mammary gland                                       797
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="methods" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="methods"><span class="header-section-number">2</span> Methods</h2>
<section id="seurat-cca" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="seurat-cca"><span class="header-section-number">2.1</span> Seurat CCA</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-cca.rds"</span>)))) {</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"batch"</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)) {</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>      obj.list[[i]] <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj.list[[i]], <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>      obj.list[[i]] <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(obj.list[[i]],</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">selection.method =</span> <span class="st">"vst"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">nfeatures =</span> <span class="dv">2000</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    panchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(<span class="at">object.list =</span> obj.list)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> panchors, <span class="at">k.weight =</span> <span class="dv">30</span>) <span class="sc">|&gt;</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunPCA</span>(<span class="at">assay =</span> <span class="st">"integrated"</span>, <span class="at">npcs =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>) <span class="sc">|&gt;</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-cca.rds"</span>)))</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-cca.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-cca.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-cca.rds"</span>)))</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-cca.rds"</span>)))</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
24593 features across 19687 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 2 layers present: data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="seurat-cca-sct" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="seurat-cca-sct"><span class="header-section-number">2.2</span> Seurat CCA SCT</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-cca-sct.rds"</span>)))) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"batch"</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)) {</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>      obj.list[[i]] <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(obj.list[[i]], <span class="at">return.only.var.genes =</span> F)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    pfeatures <span class="ot">&lt;-</span> <span class="fu">SelectIntegrationFeatures</span>(obj.list, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">PrepSCTIntegration</span>(obj.list, <span class="at">anchor.features =</span> pfeatures)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    panchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(<span class="at">object.list =</span> obj.list, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>, <span class="at">anchor.features =</span> pfeatures)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> panchors, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>, <span class="at">k.weight =</span> <span class="dv">30</span>)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">DefaultAssay</span>(obj1) <span class="ot">&lt;-</span> <span class="st">"integrated"</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(obj1, <span class="at">nps =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(obj1, <span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-cca-sct.rds"</span>)))</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-cca-sct.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-cca-sct.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-cca-sct.rds"</span>)))</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-cca-sct.rds"</span>)))</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1
[1] 2
[1] 3
[1] 4
[1] 5
[1] 6
[1] 7
Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
46742 features across 19687 samples within 3 assays 
Active assay: integrated (2000 features, 2000 variable features)
 2 layers present: data, scale.data
 2 other assays present: RNA, SCT
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="seurat-rpca" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="seurat-rpca"><span class="header-section-number">2.3</span> Seurat RPCA</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-rpca.rds"</span>)))) {</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"batch"</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> obj.list, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>      x <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(x)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>      x <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(x, <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    pfeatures <span class="ot">&lt;-</span> <span class="fu">SelectIntegrationFeatures</span>(<span class="at">object.list =</span> obj.list)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    obj.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> obj.list, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>      x <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(x, <span class="at">features =</span> pfeatures, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>      x <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(x, <span class="at">features =</span> pfeatures, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">seed.use =</span> <span class="dv">123</span>, <span class="at">npcs =</span> <span class="dv">30</span>)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>    panchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(<span class="at">object.list =</span> obj.list, <span class="at">anchor.features =</span> pfeatures, <span class="at">reduction =</span> <span class="st">"rpca"</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> panchors, <span class="at">k.weight =</span> <span class="dv">30</span>)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj1 <span class="sc">|&gt;</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunPCA</span>(<span class="at">nps =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>) <span class="sc">|&gt;</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-rpca.rds"</span>)))</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-rpca.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-rpca.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-rpca.rds"</span>)))</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-rpca.rds"</span>)))</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
24593 features across 19687 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 2 layers present: data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="harmony" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="harmony"><span class="header-section-number">2.4</span> Harmony</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-harmony.rds"</span>)))) {</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(harmony)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">FindVariableFeatures</span>() <span class="sc">|&gt;</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunPCA</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">nps =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>) <span class="sc">|&gt;</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunHarmony</span>(<span class="at">group.by.vars =</span> <span class="st">"batch"</span>) <span class="sc">|&gt;</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-harmony.rds"</span>)))</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-harmony.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-harmony.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-harmony.rds"</span>)))</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-harmony.rds"</span>)))</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 3 dimensional reductions calculated: pca, harmony, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1   D2   D3
  endothelial cell of lymphatic vessel                         26   15   11
  basal-myoepithelial cell of mammary gland                   232   60  583
  luminal adaptive secretory precursor cell of mammary gland 1264  460 1109
  luminal hormone-sensing cell of mammary gland               777  329  809
                                                            donor_id
cell_type                                                      D4   D5  D11
  endothelial cell of lymphatic vessel                         10    4   19
  basal-myoepithelial cell of mammary gland                   174  326 2305
  luminal adaptive secretory precursor cell of mammary gland  842 1193 2560
  luminal hormone-sensing cell of mammary gland               744  719  116
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  endothelial cell of lymphatic vessel                                             15
  basal-myoepithelial cell of mammary gland                                       797
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="liger" class="level3" data-number="2.5">
<h3 data-number="2.5" class="anchored" data-anchor-id="liger"><span class="header-section-number">2.5</span> Liger</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-liger.rds"</span>)))) {</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(rliger)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>    obj1[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(obj1[[<span class="st">"RNA"</span>]], <span class="at">f =</span> obj1<span class="sc">$</span>batch)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj1 <span class="sc">|&gt;</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">normalize</span>() <span class="sc">|&gt;</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">selectGenes</span>() <span class="sc">|&gt;</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scaleNotCenter</span>() <span class="sc">|&gt;</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">runINMF</span>(<span class="at">k =</span> <span class="dv">20</span>) <span class="sc">|&gt;</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">quantileNorm</span>(<span class="at">nNeighbors =</span> <span class="dv">5</span>) <span class="sc">|&gt;</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">reduction =</span> <span class="st">"inmf"</span>, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-liger.rds"</span>)))</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-liger.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-liger.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-liger.rds"</span>)))</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-liger.rds"</span>)))</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 13556 variable features)
 21 layers present: counts.pooled_D9,D7,D8,D10,D6, counts.D2, counts.D3, counts.D4, counts.D5, counts.D1, counts.D11, ligerNormData.pooled_D9,D7,D8,D10,D6, ligerNormData.D2, ligerNormData.D3, ligerNormData.D4, ligerNormData.D5, ligerNormData.D1, ligerNormData.D11, ligerScaleData.pooled_D9,D7,D8,D10,D6, ligerScaleData.D2, ligerScaleData.D3, ligerScaleData.D4, ligerScaleData.D5, ligerScaleData.D1, ligerScaleData.D11
 3 dimensional reductions calculated: inmf, inmfNorm, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1   D2   D3
  endothelial cell of lymphatic vessel                         26   15   11
  basal-myoepithelial cell of mammary gland                   232   60  583
  luminal adaptive secretory precursor cell of mammary gland 1264  460 1109
  luminal hormone-sensing cell of mammary gland               777  329  809
                                                            donor_id
cell_type                                                      D4   D5  D11
  endothelial cell of lymphatic vessel                         10    4   19
  basal-myoepithelial cell of mammary gland                   174  326 2305
  luminal adaptive secretory precursor cell of mammary gland  842 1193 2560
  luminal hormone-sensing cell of mammary gland               744  719  116
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  endothelial cell of lymphatic vessel                                             15
  basal-myoepithelial cell of mammary gland                                       797
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="fastmnn" class="level3" data-number="2.6">
<h3 data-number="2.6" class="anchored" data-anchor-id="fastmnn"><span class="header-section-number">2.6</span> FastMNN</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-fastmnn.rds"</span>)))) {</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(batchelor)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">FindVariableFeatures</span>()</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">RunFastMNN</span>(<span class="at">object.list =</span> <span class="fu">SplitObject</span>(obj1, <span class="at">split.by =</span> <span class="st">"batch"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"mnn"</span>, <span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-fastmnn.rds"</span>)))</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-fastmnn.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-fastmnn.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-fastmnn.rds"</span>)))</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-fastmnn.rds"</span>)))</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
24593 features across 19687 samples within 2 assays 
Active assay: RNA (22593 features, 2000 variable features)
 14 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, data.1, data.2, data.3, data.4, data.5, data.6, data.7
 1 other assay present: mnn.reconstructed
 2 dimensional reductions calculated: mnn, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="stacas" class="level3" data-number="2.7">
<h3 data-number="2.7" class="anchored" data-anchor-id="stacas"><span class="header-section-number">2.7</span> STACAS</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-stacas.rds"</span>)))) {</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(STACAS)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">SplitObject</span>(<span class="at">split.by =</span> <span class="st">"batch"</span>) <span class="sc">|&gt;</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">Run.STACAS</span>(<span class="at">min.sample.size =</span> <span class="dv">20</span>, <span class="at">k.weight =</span> <span class="dv">30</span>) <span class="sc">|&gt;</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-stacas.rds"</span>)))</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-stacas.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-stacas.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-stacas.rds"</span>)))</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-stacas.rds"</span>)))</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
23593 features across 19687 samples within 2 assays 
Active assay: integrated (1000 features, 1000 variable features)
 2 layers present: data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="stacas-grp" class="level3" data-number="2.8">
<h3 data-number="2.8" class="anchored" data-anchor-id="stacas-grp"><span class="header-section-number">2.8</span> STACAS Grp</h3>
<p>STACAS run in semi-supervised mode where celltypes are specified.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-stacas-grp.rds"</span>)))) {</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(STACAS)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">SplitObject</span>(<span class="at">split.by =</span> <span class="st">"batch"</span>) <span class="sc">|&gt;</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">Run.STACAS</span>(<span class="at">min.sample.size =</span> <span class="dv">20</span>, <span class="at">cell.labels =</span> grp, <span class="at">k.weight =</span> <span class="dv">30</span>) <span class="sc">|&gt;</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">RunUMAP</span>(<span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-stacas-grp.rds"</span>)))</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-stacas-grp.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-stacas-grp.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-stacas-grp.rds"</span>)))</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-stacas-grp.rds"</span>)))</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
23593 features across 19687 samples within 2 assays 
Active assay: integrated (1000 features, 1000 variable features)
 2 layers present: data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="conos" class="level3" data-number="2.9">
<h3 data-number="2.9" class="anchored" data-anchor-id="conos"><span class="header-section-number">2.9</span> Conos</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-conos.rds"</span>)))) {</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(conos)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>    obj.panel <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"batch"</span>)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.panel)) {</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>      obj.panel[[i]] <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj.panel[[i]]) <span class="sc">|&gt;</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">FindVariableFeatures</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">RunPCA</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">nps =</span> <span class="dv">30</span>, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>    obj.con <span class="ot">&lt;-</span> Conos<span class="sc">$</span><span class="fu">new</span>(obj.panel)</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(obj.panel)</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>    obj.con<span class="sc">$</span><span class="fu">buildGraph</span>(<span class="at">k =</span> <span class="dv">15</span>, <span class="at">k.self =</span> <span class="dv">5</span>, <span class="at">space =</span> <span class="st">"PCA"</span>, <span class="at">ncomps =</span> <span class="dv">20</span>, <span class="at">n.odgenes =</span> <span class="dv">2000</span>, <span class="at">matching.method =</span> <span class="st">"mNN"</span>, <span class="at">metric =</span> <span class="st">"angular"</span>, <span class="at">score.component.variance =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>    obj.con<span class="sc">$</span><span class="fu">findCommunities</span>()</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>    obj.con<span class="sc">$</span><span class="fu">embedGraph</span>()</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">as.Seurat</span>(obj.con)</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>    obj1[[<span class="st">"umap"</span>]] <span class="ot">&lt;-</span> obj1[[<span class="st">"largeVis"</span>]]</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-conos.rds"</span>)))</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-conos.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-conos.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-conos.rds"</span>)))</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-conos.rds"</span>)))</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 2000 variable features)
 21 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, data.1, scale.data.1, data.2, scale.data.2, data.3, scale.data.3, data.4, scale.data.4, data.5, scale.data.5, data.6, scale.data.6, data.7, scale.data.7
 2 dimensional reductions calculated: largeVis, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1  D11   D2
  basal-myoepithelial cell of mammary gland                   232 2305   60
  endothelial cell of lymphatic vessel                         26   19   15
  luminal adaptive secretory precursor cell of mammary gland 1264 2560  460
  luminal hormone-sensing cell of mammary gland               777  116  329
                                                            donor_id
cell_type                                                      D3   D4   D5
  basal-myoepithelial cell of mammary gland                   583  174  326
  endothelial cell of lymphatic vessel                         11   10    4
  luminal adaptive secretory precursor cell of mammary gland 1109  842 1193
  luminal hormone-sensing cell of mammary gland               809  744  719
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  basal-myoepithelial cell of mammary gland                                       797
  endothelial cell of lymphatic vessel                                             15
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="scanorama" class="level3" data-number="2.10">
<h3 data-number="2.10" class="anchored" data-anchor-id="scanorama"><span class="header-section-number">2.10</span> Scanorama</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-scanorama.rds"</span>)))) {</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sys.setenv(PYTHONPATH="/home/rstudio/.local/lib/python3.10/site-packages")</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">Sys.setenv</span>(<span class="at">PYTHONPATH=</span><span class="st">"/usr/lib/python3.10/site-packages"</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(reticulate)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#reticulate::use_condaenv("/crex/proj/snic2022-22-328/nobackup/roy/integration/conda/scanpy")</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">gc</span>()</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    scanorama <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scanorama"</span>)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span> <span class="fu">NormalizeData</span>()</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    obj1.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj1, <span class="at">split.by =</span> <span class="st">"batch"</span>)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    assaylist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    genelist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj1.list))</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>      assaylist[[i]] <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">as.matrix</span>(<span class="fu">GetAssayData</span>(obj1.list[[i]], <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">layer =</span> <span class="st">"data"</span>)))</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>      genelist[[i]] <span class="ot">&lt;-</span> <span class="fu">rownames</span>(obj1.list[[i]])</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>    mat_integrated <span class="ot">&lt;-</span> scanorama<span class="sc">$</span><span class="fu">integrate</span>(assaylist, genelist)</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    dr_scanorama <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, mat_integrated[[<span class="dv">1</span>]])</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(dr_scanorama) <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, <span class="fu">lapply</span>(assaylist, rownames))</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    dr_scanorama <span class="ot">&lt;-</span> dr_scanorama[<span class="fu">colnames</span>(obj1), ]</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>    obj2 <span class="ot">&lt;-</span> obj</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    obj2[[<span class="st">"scanorama"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateDimReducObject</span>(dr_scanorama, <span class="at">key =</span> <span class="st">"SCANORAMA_"</span>)</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>    obj2 <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(obj2, <span class="at">reduction =</span> <span class="st">"scanorama"</span>, <span class="at">dims =</span> dims_umap, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>  obj2</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj2, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-scanorama.rds"</span>)))</span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-scanorama.png"</span>)), <span class="fu">plt</span>(obj2, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-scanorama.png"</span>)), <span class="fu">plt</span>(obj2, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-scanorama.rds"</span>)))</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj2, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-scanorama.rds"</span>)))</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1,obj2)</span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Found 22593 genes among all datasets
[[0.         0.51041667 0.1462058  0.08074534 0.02005348 0.00347373
  0.60316609]
 [0.         0.         0.86574074 0.33333333 0.07638889 0.18518519
  0.08449074]
 [0.         0.         0.         0.59992054 0.21771951 0.24076281
  0.09574891]
 [0.         0.         0.         0.         0.73395722 0.55223038
  0.02879729]
 [0.         0.         0.         0.         0.         0.91755793
  0.01381462]
 [0.         0.         0.         0.         0.         0.
  0.07468519]
 [0.         0.         0.         0.         0.         0.
  0.        ]]
Processing datasets (4, 5)
Processing datasets (1, 2)
Processing datasets (3, 4)
Processing datasets (0, 6)
Processing datasets (2, 3)
Processing datasets (3, 5)
Processing datasets (0, 1)
Processing datasets (1, 3)
Processing datasets (2, 5)
Processing datasets (2, 4)
Processing datasets (1, 5)
Processing datasets (0, 2)
Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 0 variable features)
 1 layer present: counts
 2 dimensional reductions calculated: scanorama, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1   D2   D3
  endothelial cell of lymphatic vessel                         26   15   11
  basal-myoepithelial cell of mammary gland                   232   60  583
  luminal adaptive secretory precursor cell of mammary gland 1264  460 1109
  luminal hormone-sensing cell of mammary gland               777  329  809
                                                            donor_id
cell_type                                                      D4   D5  D11
  endothelial cell of lymphatic vessel                         10    4   19
  basal-myoepithelial cell of mammary gland                   174  326 2305
  luminal adaptive secretory precursor cell of mammary gland  842 1193 2560
  luminal hormone-sensing cell of mammary gland               744  719  116
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  endothelial cell of lymphatic vessel                                             15
  basal-myoepithelial cell of mammary gland                                       797
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
</section>
<section id="scvi" class="level3" data-number="2.11">
<h3 data-number="2.11" class="anchored" data-anchor-id="scvi"><span class="header-section-number">2.11</span> scVI</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-scvi.rds"</span>)))) {</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gc</span>()</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  start_time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  memtime <span class="ot">&lt;-</span> <span class="fu">peakRAM</span>({</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sys.setenv(PYTHONPATH="/home/rstudio/.local/lib/python3.10/site-packages")</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">Sys.setenv</span>(<span class="at">PYTHONPATH=</span><span class="st">"/usr/lib/python3.10/site-packages"</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(reticulate)</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(sceasy)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#use_condaenv("/crex/proj/snic2022-22-328/nobackup/roy/integration/conda/scanpy")</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>    sc <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scanpy"</span>, <span class="at">convert =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>    scvi <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scvi"</span>, <span class="at">convert =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Trainer(accelerator='gpu')</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> obj <span class="sc">|&gt;</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">NormalizeData</span>(<span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>) <span class="sc">|&gt;</span></span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">FindVariableFeatures</span>(<span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># convert v5 assay to v3/4 assay since sceasy is outdated</span></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>    obj1[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">as</span>(obj1[[<span class="st">"RNA"</span>]], <span class="st">"Assay"</span>)</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>    adata <span class="ot">&lt;-</span> sceasy<span class="sc">::</span><span class="fu">convertFormat</span>(obj1, <span class="at">from =</span> <span class="st">"seurat"</span>, <span class="at">to =</span> <span class="st">"anndata"</span>, <span class="at">main_layer =</span> <span class="st">"counts"</span>, <span class="at">drop_single_values =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(adata)</span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># run setup_anndata, use column stim for batch</span></span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>    scvi<span class="sc">$</span>model<span class="sc">$</span>SCVI<span class="sc">$</span><span class="fu">setup_anndata</span>(adata, <span class="at">batch_key =</span> <span class="st">"batch"</span>)</span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>    <span class="co"># scvi$settings.dl_num_workers &lt;- as.integer(8)</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># create the model</span></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a>    model <span class="ot">&lt;-</span> scvi<span class="sc">$</span>model<span class="sc">$</span><span class="fu">SCVI</span>(adata)</span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># train the model</span></span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># model$train()</span></span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># to specify the number of epochs when training:</span></span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>    model<span class="sc">$</span><span class="fu">train</span>(<span class="at">max_epochs =</span> <span class="fu">as.integer</span>(<span class="dv">100</span>))</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a>    latent <span class="ot">&lt;-</span> model<span class="sc">$</span><span class="fu">get_latent_representation</span>()</span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>    latent <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(latent)</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(latent) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(obj1)</span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a>    obj1[[<span class="st">"scvi"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateDimReducObject</span>(<span class="at">embeddings =</span> latent, <span class="at">key =</span> <span class="st">"scvi_"</span>, <span class="at">assay =</span> <span class="fu">DefaultAssay</span>(obj1))</span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a>    obj1 <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(obj1, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(obj1<span class="sc">@</span>reductions<span class="sc">$</span>scvi<span class="sc">@</span>cell.embeddings), <span class="at">reduction =</span> <span class="st">"scvi"</span>, <span class="at">n.components =</span> <span class="dv">2</span>, <span class="at">seed.use =</span> <span class="dv">123</span>)</span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a>  obj1</span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> <span class="fu">fn_getmetrics</span>(obj1, <span class="at">batch =</span> batch, <span class="at">grp =</span> grp, <span class="at">metrics_ilp =</span> metrics_ilp, <span class="at">metrics_red =</span> metrics_red)</span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(metrics, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-metrics-scvi.rds"</span>)))</span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(<span class="fu">as.numeric</span>(<span class="fu">difftime</span>(<span class="fu">Sys.time</span>(), start_time, <span class="at">units =</span> <span class="st">"mins"</span>)), <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-time-scvi.rds"</span>)))</span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, grp, <span class="st">"-scvi.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp))</span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-umap-"</span>, batch, <span class="st">"-scvi.png"</span>)), <span class="fu">plt</span>(obj1, <span class="at">grp =</span> batch, <span class="at">cols =</span> cols_batch))</span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(memtime, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-memtime-scvi.rds"</span>)))</span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj1, <span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-seurat-scvi.rds"</span>)))</span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(obj1)</span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>AnnData object with n_obs  n_vars = 31696  22593
    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'mapped_reference_annotation', 'donor_id', 'self_reported_ethnicity_ontology_term_id', 'family_history_breast_cancer', 'donor_living_at_sample_collection', 'organism_ontology_term_id', 'tyrer_cuzick_lifetime_risk', 'donor_times_pregnant', 'sample_uuid', 'sample_preservation_method', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'sample_derivation_process', 'sample_source', 'donor_BMI_at_collection', 'tissue_type', 'suspension_derivation_process', 'suspension_dissociation_reagent', 'suspension_dissociation_time', 'suspension_percent_cell_viability', 'suspension_uuid', 'suspension_type', 'tissue_handling_interval', 'library_uuid', 'assay_ontology_term_id', 'sequencing_platform', 'is_primary_data', 'cell_type_ontology_term_id', 'author_cell_type', 'disease_ontology_term_id', 'sex_ontology_term_id', 'percent.mito', 'seurat_clusters', 'sample_id', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'batch'
    var: 'vf_vst_counts_mean', 'vf_vst_counts_variance', 'vf_vst_counts_variance.expected', 'vf_vst_counts_variance.standardized', 'vf_vst_counts_variable', 'vf_vst_counts_rank', 'var.features', 'var.features.rank'

Training:   0%|          | 0/100 [00:00&lt;?, ?it/s]
Epoch 1/100:   0%|          | 0/100 [00:00&lt;?, ?it/s]
Epoch 1/100:   1%|          | 1/100 [01:34&lt;2:36:15, 94.70s/it]
Epoch 1/100:   1%|          | 1/100 [01:34&lt;2:36:15, 94.70s/it, v_num=1, train_loss_step=5.31e+3, train_loss_epoch=5.66e+3]
Epoch 2/100:   1%|          | 1/100 [01:34&lt;2:36:15, 94.70s/it, v_num=1, train_loss_step=5.31e+3, train_loss_epoch=5.66e+3]
Epoch 2/100:   2%|         | 2/100 [03:10&lt;2:35:35, 95.26s/it, v_num=1, train_loss_step=5.31e+3, train_loss_epoch=5.66e+3]
Epoch 2/100:   2%|         | 2/100 [03:10&lt;2:35:35, 95.26s/it, v_num=1, train_loss_step=4.99e+3, train_loss_epoch=5.2e+3] 
Epoch 3/100:   2%|         | 2/100 [03:10&lt;2:35:35, 95.26s/it, v_num=1, train_loss_step=4.99e+3, train_loss_epoch=5.2e+3]
Epoch 3/100:   3%|         | 3/100 [04:45&lt;2:34:05, 95.32s/it, v_num=1, train_loss_step=4.99e+3, train_loss_epoch=5.2e+3]
Epoch 3/100:   3%|         | 3/100 [04:45&lt;2:34:05, 95.32s/it, v_num=1, train_loss_step=5.15e+3, train_loss_epoch=5.1e+3]
Epoch 4/100:   3%|         | 3/100 [04:45&lt;2:34:05, 95.32s/it, v_num=1, train_loss_step=5.15e+3, train_loss_epoch=5.1e+3]
Epoch 4/100:   4%|         | 4/100 [06:21&lt;2:32:32, 95.34s/it, v_num=1, train_loss_step=5.15e+3, train_loss_epoch=5.1e+3]
Epoch 4/100:   4%|         | 4/100 [06:21&lt;2:32:32, 95.34s/it, v_num=1, train_loss_step=5.01e+3, train_loss_epoch=5.04e+3]
Epoch 5/100:   4%|         | 4/100 [06:21&lt;2:32:32, 95.34s/it, v_num=1, train_loss_step=5.01e+3, train_loss_epoch=5.04e+3]
Epoch 5/100:   5%|         | 5/100 [07:56&lt;2:30:45, 95.21s/it, v_num=1, train_loss_step=5.01e+3, train_loss_epoch=5.04e+3]
Epoch 5/100:   5%|         | 5/100 [07:56&lt;2:30:45, 95.21s/it, v_num=1, train_loss_step=5.03e+3, train_loss_epoch=4.99e+3]
Epoch 6/100:   5%|         | 5/100 [07:56&lt;2:30:45, 95.21s/it, v_num=1, train_loss_step=5.03e+3, train_loss_epoch=4.99e+3]
Epoch 6/100:   6%|         | 6/100 [09:30&lt;2:28:50, 95.01s/it, v_num=1, train_loss_step=5.03e+3, train_loss_epoch=4.99e+3]
Epoch 6/100:   6%|         | 6/100 [09:30&lt;2:28:50, 95.01s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.95e+3]
Epoch 7/100:   6%|         | 6/100 [09:30&lt;2:28:50, 95.01s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.95e+3]
Epoch 7/100:   7%|         | 7/100 [11:07&lt;2:28:00, 95.49s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.95e+3]
Epoch 7/100:   7%|         | 7/100 [11:07&lt;2:28:00, 95.49s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.92e+3]
Epoch 8/100:   7%|         | 7/100 [11:07&lt;2:28:00, 95.49s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.92e+3]
Epoch 8/100:   8%|         | 8/100 [12:44&lt;2:27:04, 95.92s/it, v_num=1, train_loss_step=4.94e+3, train_loss_epoch=4.92e+3]
Epoch 8/100:   8%|         | 8/100 [12:44&lt;2:27:04, 95.92s/it, v_num=1, train_loss_step=5.05e+3, train_loss_epoch=4.9e+3] 
Epoch 9/100:   8%|         | 8/100 [12:44&lt;2:27:04, 95.92s/it, v_num=1, train_loss_step=5.05e+3, train_loss_epoch=4.9e+3]
Epoch 9/100:   9%|         | 9/100 [14:21&lt;2:26:01, 96.28s/it, v_num=1, train_loss_step=5.05e+3, train_loss_epoch=4.9e+3]
Epoch 9/100:   9%|         | 9/100 [14:21&lt;2:26:01, 96.28s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.87e+3]
Epoch 10/100:   9%|         | 9/100 [14:21&lt;2:26:01, 96.28s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.87e+3]
Epoch 10/100:  10%|         | 10/100 [15:58&lt;2:24:57, 96.63s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.87e+3]
Epoch 10/100:  10%|         | 10/100 [15:58&lt;2:24:57, 96.63s/it, v_num=1, train_loss_step=4.85e+3, train_loss_epoch=4.85e+3]
Epoch 11/100:  10%|         | 10/100 [15:58&lt;2:24:57, 96.63s/it, v_num=1, train_loss_step=4.85e+3, train_loss_epoch=4.85e+3]
Epoch 11/100:  11%|         | 11/100 [17:36&lt;2:23:56, 97.04s/it, v_num=1, train_loss_step=4.85e+3, train_loss_epoch=4.85e+3]
Epoch 11/100:  11%|         | 11/100 [17:36&lt;2:23:56, 97.04s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.83e+3]
Epoch 12/100:  11%|         | 11/100 [17:36&lt;2:23:56, 97.04s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.83e+3]
Epoch 12/100:  12%|        | 12/100 [19:13&lt;2:22:30, 97.16s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.83e+3]
Epoch 12/100:  12%|        | 12/100 [19:13&lt;2:22:30, 97.16s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.82e+3]
Epoch 13/100:  12%|        | 12/100 [19:13&lt;2:22:30, 97.16s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.82e+3]
Epoch 13/100:  13%|        | 13/100 [20:50&lt;2:20:46, 97.09s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.82e+3]
Epoch 13/100:  13%|        | 13/100 [20:50&lt;2:20:46, 97.09s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.81e+3]
Epoch 14/100:  13%|        | 13/100 [20:50&lt;2:20:46, 97.09s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.81e+3]
Epoch 14/100:  14%|        | 14/100 [22:27&lt;2:19:04, 97.03s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.81e+3]
Epoch 14/100:  14%|        | 14/100 [22:27&lt;2:19:04, 97.03s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.79e+3]
Epoch 15/100:  14%|        | 14/100 [22:27&lt;2:19:04, 97.03s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.79e+3]
Epoch 15/100:  15%|        | 15/100 [24:04&lt;2:17:29, 97.05s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.79e+3]
Epoch 15/100:  15%|        | 15/100 [24:04&lt;2:17:29, 97.05s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.78e+3]
Epoch 16/100:  15%|        | 15/100 [24:04&lt;2:17:29, 97.05s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.78e+3]
Epoch 16/100:  16%|        | 16/100 [25:42&lt;2:16:04, 97.20s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.78e+3]
Epoch 16/100:  16%|        | 16/100 [25:42&lt;2:16:04, 97.20s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.77e+3]
Epoch 17/100:  16%|        | 16/100 [25:42&lt;2:16:04, 97.20s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.77e+3]
Epoch 17/100:  17%|        | 17/100 [27:21&lt;2:15:06, 97.67s/it, v_num=1, train_loss_step=4.87e+3, train_loss_epoch=4.77e+3]
Epoch 17/100:  17%|        | 17/100 [27:21&lt;2:15:06, 97.67s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.77e+3]
Epoch 18/100:  17%|        | 17/100 [27:21&lt;2:15:06, 97.67s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.77e+3]
Epoch 18/100:  18%|        | 18/100 [28:59&lt;2:13:39, 97.80s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.77e+3]
Epoch 18/100:  18%|        | 18/100 [28:59&lt;2:13:39, 97.80s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.76e+3]
Epoch 19/100:  18%|        | 18/100 [28:59&lt;2:13:39, 97.80s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.76e+3]
Epoch 19/100:  19%|        | 19/100 [30:38&lt;2:12:30, 98.16s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.76e+3]
Epoch 19/100:  19%|        | 19/100 [30:38&lt;2:12:30, 98.16s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.75e+3]
Epoch 20/100:  19%|        | 19/100 [30:38&lt;2:12:30, 98.16s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.75e+3]
Epoch 20/100:  20%|        | 20/100 [32:17&lt;2:11:12, 98.40s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.75e+3]
Epoch 20/100:  20%|        | 20/100 [32:17&lt;2:11:12, 98.40s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.74e+3]
Epoch 21/100:  20%|        | 20/100 [32:17&lt;2:11:12, 98.40s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.74e+3]
Epoch 21/100:  21%|        | 21/100 [33:53&lt;2:08:42, 97.76s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.74e+3]
Epoch 21/100:  21%|        | 21/100 [33:53&lt;2:08:42, 97.76s/it, v_num=1, train_loss_step=4.81e+3, train_loss_epoch=4.74e+3]
Epoch 22/100:  21%|        | 21/100 [33:53&lt;2:08:42, 97.76s/it, v_num=1, train_loss_step=4.81e+3, train_loss_epoch=4.74e+3]
Epoch 22/100:  22%|       | 22/100 [35:27&lt;2:05:46, 96.75s/it, v_num=1, train_loss_step=4.81e+3, train_loss_epoch=4.74e+3]
Epoch 22/100:  22%|       | 22/100 [35:27&lt;2:05:46, 96.75s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.73e+3]
Epoch 23/100:  22%|       | 22/100 [35:27&lt;2:05:46, 96.75s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.73e+3]
Epoch 23/100:  23%|       | 23/100 [37:02&lt;2:03:16, 96.06s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.73e+3]
Epoch 23/100:  23%|       | 23/100 [37:02&lt;2:03:16, 96.06s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.73e+3]
Epoch 24/100:  23%|       | 23/100 [37:02&lt;2:03:16, 96.06s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.73e+3]
Epoch 24/100:  24%|       | 24/100 [38:36&lt;2:00:59, 95.52s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.73e+3]
Epoch 24/100:  24%|       | 24/100 [38:36&lt;2:00:59, 95.52s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 25/100:  24%|       | 24/100 [38:36&lt;2:00:59, 95.52s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 25/100:  25%|       | 25/100 [40:11&lt;1:59:00, 95.21s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 25/100:  25%|       | 25/100 [40:11&lt;1:59:00, 95.21s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.72e+3]
Epoch 26/100:  25%|       | 25/100 [40:11&lt;1:59:00, 95.21s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.72e+3]
Epoch 26/100:  26%|       | 26/100 [41:45&lt;1:57:17, 95.11s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.72e+3]
Epoch 26/100:  26%|       | 26/100 [41:45&lt;1:57:17, 95.11s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 27/100:  26%|       | 26/100 [41:45&lt;1:57:17, 95.11s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 27/100:  27%|       | 27/100 [43:20&lt;1:55:28, 94.91s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.72e+3]
Epoch 27/100:  27%|       | 27/100 [43:20&lt;1:55:28, 94.91s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.71e+3]
Epoch 28/100:  27%|       | 27/100 [43:20&lt;1:55:28, 94.91s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.71e+3]
Epoch 28/100:  28%|       | 28/100 [44:54&lt;1:53:44, 94.78s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.71e+3]
Epoch 28/100:  28%|       | 28/100 [44:54&lt;1:53:44, 94.78s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.71e+3]
Epoch 29/100:  28%|       | 28/100 [44:54&lt;1:53:44, 94.78s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.71e+3]
Epoch 29/100:  29%|       | 29/100 [46:29&lt;1:51:57, 94.61s/it, v_num=1, train_loss_step=4.83e+3, train_loss_epoch=4.71e+3]
Epoch 29/100:  29%|       | 29/100 [46:29&lt;1:51:57, 94.61s/it, v_num=1, train_loss_step=4.78e+3, train_loss_epoch=4.71e+3]
Epoch 30/100:  29%|       | 29/100 [46:29&lt;1:51:57, 94.61s/it, v_num=1, train_loss_step=4.78e+3, train_loss_epoch=4.71e+3]
Epoch 30/100:  30%|       | 30/100 [48:03&lt;1:50:20, 94.58s/it, v_num=1, train_loss_step=4.78e+3, train_loss_epoch=4.71e+3]
Epoch 30/100:  30%|       | 30/100 [48:03&lt;1:50:20, 94.58s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.7e+3] 
Epoch 31/100:  30%|       | 30/100 [48:03&lt;1:50:20, 94.58s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.7e+3]
Epoch 31/100:  31%|       | 31/100 [49:37&lt;1:48:42, 94.52s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.7e+3]
Epoch 31/100:  31%|       | 31/100 [49:37&lt;1:48:42, 94.52s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.7e+3]
Epoch 32/100:  31%|       | 31/100 [49:37&lt;1:48:42, 94.52s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.7e+3]
Epoch 32/100:  32%|      | 32/100 [51:12&lt;1:47:06, 94.51s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.7e+3]
Epoch 32/100:  32%|      | 32/100 [51:12&lt;1:47:06, 94.51s/it, v_num=1, train_loss_step=4.79e+3, train_loss_epoch=4.7e+3]
Epoch 33/100:  32%|      | 32/100 [51:12&lt;1:47:06, 94.51s/it, v_num=1, train_loss_step=4.79e+3, train_loss_epoch=4.7e+3]
Epoch 33/100:  33%|      | 33/100 [52:47&lt;1:45:43, 94.67s/it, v_num=1, train_loss_step=4.79e+3, train_loss_epoch=4.7e+3]
Epoch 33/100:  33%|      | 33/100 [52:47&lt;1:45:43, 94.67s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.7e+3]
Epoch 34/100:  33%|      | 33/100 [52:47&lt;1:45:43, 94.67s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.7e+3]
Epoch 34/100:  34%|      | 34/100 [54:21&lt;1:44:02, 94.58s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.7e+3]
Epoch 34/100:  34%|      | 34/100 [54:21&lt;1:44:02, 94.58s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.7e+3]
Epoch 35/100:  34%|      | 34/100 [54:21&lt;1:44:02, 94.58s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.7e+3]
Epoch 35/100:  35%|      | 35/100 [55:55&lt;1:42:15, 94.39s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.7e+3]
Epoch 35/100:  35%|      | 35/100 [55:55&lt;1:42:15, 94.39s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 36/100:  35%|      | 35/100 [55:55&lt;1:42:15, 94.39s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 36/100:  36%|      | 36/100 [57:30&lt;1:40:40, 94.39s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 36/100:  36%|      | 36/100 [57:30&lt;1:40:40, 94.39s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 37/100:  36%|      | 36/100 [57:30&lt;1:40:40, 94.39s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 37/100:  37%|      | 37/100 [59:04&lt;1:39:00, 94.30s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.69e+3]
Epoch 37/100:  37%|      | 37/100 [59:04&lt;1:39:00, 94.30s/it, v_num=1, train_loss_step=4.54e+3, train_loss_epoch=4.69e+3]
Epoch 38/100:  37%|      | 37/100 [59:04&lt;1:39:00, 94.30s/it, v_num=1, train_loss_step=4.54e+3, train_loss_epoch=4.69e+3]
Epoch 38/100:  38%|      | 38/100 [1:00:38&lt;1:37:30, 94.36s/it, v_num=1, train_loss_step=4.54e+3, train_loss_epoch=4.69e+3]
Epoch 38/100:  38%|      | 38/100 [1:00:38&lt;1:37:30, 94.36s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.69e+3]
Epoch 39/100:  38%|      | 38/100 [1:00:38&lt;1:37:30, 94.36s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.69e+3]
Epoch 39/100:  39%|      | 39/100 [1:02:12&lt;1:35:48, 94.24s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.69e+3]
Epoch 39/100:  39%|      | 39/100 [1:02:12&lt;1:35:48, 94.24s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.69e+3]
Epoch 40/100:  39%|      | 39/100 [1:02:12&lt;1:35:48, 94.24s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.69e+3]
Epoch 40/100:  40%|      | 40/100 [1:03:46&lt;1:34:02, 94.04s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.69e+3]
Epoch 40/100:  40%|      | 40/100 [1:03:46&lt;1:34:02, 94.04s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.69e+3]
Epoch 41/100:  40%|      | 40/100 [1:03:46&lt;1:34:02, 94.04s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.69e+3]
Epoch 41/100:  41%|      | 41/100 [1:05:20&lt;1:32:33, 94.12s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.69e+3]
Epoch 41/100:  41%|      | 41/100 [1:05:20&lt;1:32:33, 94.12s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 42/100:  41%|      | 41/100 [1:05:20&lt;1:32:33, 94.12s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 42/100:  42%|     | 42/100 [1:06:55&lt;1:31:13, 94.37s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 42/100:  42%|     | 42/100 [1:06:55&lt;1:31:13, 94.37s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.68e+3]
Epoch 43/100:  42%|     | 42/100 [1:06:55&lt;1:31:13, 94.37s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.68e+3]
Epoch 43/100:  43%|     | 43/100 [1:08:30&lt;1:29:50, 94.58s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.68e+3]
Epoch 43/100:  43%|     | 43/100 [1:08:30&lt;1:29:50, 94.58s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.68e+3]
Epoch 44/100:  43%|     | 43/100 [1:08:30&lt;1:29:50, 94.58s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.68e+3]
Epoch 44/100:  44%|     | 44/100 [1:10:05&lt;1:28:21, 94.66s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.68e+3]
Epoch 44/100:  44%|     | 44/100 [1:10:05&lt;1:28:21, 94.66s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 45/100:  44%|     | 44/100 [1:10:05&lt;1:28:21, 94.66s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 45/100:  45%|     | 45/100 [1:11:40&lt;1:26:45, 94.64s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 45/100:  45%|     | 45/100 [1:11:40&lt;1:26:45, 94.64s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 46/100:  45%|     | 45/100 [1:11:40&lt;1:26:45, 94.64s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 46/100:  46%|     | 46/100 [1:13:15&lt;1:25:21, 94.84s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 46/100:  46%|     | 46/100 [1:13:15&lt;1:25:21, 94.84s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 47/100:  46%|     | 46/100 [1:13:15&lt;1:25:21, 94.84s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 47/100:  47%|     | 47/100 [1:14:50&lt;1:23:48, 94.88s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.68e+3]
Epoch 47/100:  47%|     | 47/100 [1:14:50&lt;1:23:48, 94.88s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 48/100:  47%|     | 47/100 [1:14:50&lt;1:23:48, 94.88s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 48/100:  48%|     | 48/100 [1:16:24&lt;1:22:00, 94.63s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.68e+3]
Epoch 48/100:  48%|     | 48/100 [1:16:24&lt;1:22:00, 94.63s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.68e+3]
Epoch 49/100:  48%|     | 48/100 [1:16:24&lt;1:22:00, 94.63s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.68e+3]
Epoch 49/100:  49%|     | 49/100 [1:17:57&lt;1:20:07, 94.26s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.68e+3]
Epoch 49/100:  49%|     | 49/100 [1:17:57&lt;1:20:07, 94.26s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.68e+3]
Epoch 50/100:  49%|     | 49/100 [1:17:57&lt;1:20:07, 94.26s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.68e+3]
Epoch 50/100:  50%|     | 50/100 [1:19:30&lt;1:18:11, 93.83s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.68e+3]
Epoch 50/100:  50%|     | 50/100 [1:19:30&lt;1:18:11, 93.83s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.68e+3]
Epoch 51/100:  50%|     | 50/100 [1:19:30&lt;1:18:11, 93.83s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.68e+3]
Epoch 51/100:  51%|     | 51/100 [1:21:03&lt;1:16:25, 93.59s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.68e+3]
Epoch 51/100:  51%|     | 51/100 [1:21:03&lt;1:16:25, 93.59s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.67e+3]
Epoch 52/100:  51%|     | 51/100 [1:21:03&lt;1:16:25, 93.59s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.67e+3]
Epoch 52/100:  52%|    | 52/100 [1:22:37&lt;1:14:52, 93.59s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.67e+3]
Epoch 52/100:  52%|    | 52/100 [1:22:37&lt;1:14:52, 93.59s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.67e+3]
Epoch 53/100:  52%|    | 52/100 [1:22:37&lt;1:14:52, 93.59s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.67e+3]
Epoch 53/100:  53%|    | 53/100 [1:24:11&lt;1:13:24, 93.70s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.67e+3]
Epoch 53/100:  53%|    | 53/100 [1:24:11&lt;1:13:24, 93.70s/it, v_num=1, train_loss_step=4.47e+3, train_loss_epoch=4.67e+3]
Epoch 54/100:  53%|    | 53/100 [1:24:11&lt;1:13:24, 93.70s/it, v_num=1, train_loss_step=4.47e+3, train_loss_epoch=4.67e+3]
Epoch 54/100:  54%|    | 54/100 [1:25:45&lt;1:11:55, 93.82s/it, v_num=1, train_loss_step=4.47e+3, train_loss_epoch=4.67e+3]
Epoch 54/100:  54%|    | 54/100 [1:25:45&lt;1:11:55, 93.82s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.67e+3]
Epoch 55/100:  54%|    | 54/100 [1:25:45&lt;1:11:55, 93.82s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.67e+3]
Epoch 55/100:  55%|    | 55/100 [1:27:19&lt;1:10:24, 93.87s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.67e+3]
Epoch 55/100:  55%|    | 55/100 [1:27:19&lt;1:10:24, 93.87s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.67e+3]
Epoch 56/100:  55%|    | 55/100 [1:27:19&lt;1:10:24, 93.87s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.67e+3]
Epoch 56/100:  56%|    | 56/100 [1:28:53&lt;1:08:56, 94.01s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.67e+3]
Epoch 56/100:  56%|    | 56/100 [1:28:53&lt;1:08:56, 94.01s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.67e+3]
Epoch 57/100:  56%|    | 56/100 [1:28:53&lt;1:08:56, 94.01s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.67e+3]
Epoch 57/100:  57%|    | 57/100 [1:30:28&lt;1:07:33, 94.27s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.67e+3]
Epoch 57/100:  57%|    | 57/100 [1:30:28&lt;1:07:33, 94.27s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.67e+3]
Epoch 58/100:  57%|    | 57/100 [1:30:28&lt;1:07:33, 94.27s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.67e+3]
Epoch 58/100:  58%|    | 58/100 [1:32:03&lt;1:06:09, 94.51s/it, v_num=1, train_loss_step=4.72e+3, train_loss_epoch=4.67e+3]
Epoch 58/100:  58%|    | 58/100 [1:32:03&lt;1:06:09, 94.51s/it, v_num=1, train_loss_step=4.49e+3, train_loss_epoch=4.67e+3]
Epoch 59/100:  58%|    | 58/100 [1:32:03&lt;1:06:09, 94.51s/it, v_num=1, train_loss_step=4.49e+3, train_loss_epoch=4.67e+3]
Epoch 59/100:  59%|    | 59/100 [1:33:38&lt;1:04:44, 94.74s/it, v_num=1, train_loss_step=4.49e+3, train_loss_epoch=4.67e+3]
Epoch 59/100:  59%|    | 59/100 [1:33:38&lt;1:04:44, 94.74s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 60/100:  59%|    | 59/100 [1:33:38&lt;1:04:44, 94.74s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 60/100:  60%|    | 60/100 [1:35:13&lt;1:03:13, 94.83s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 60/100:  60%|    | 60/100 [1:35:13&lt;1:03:13, 94.83s/it, v_num=1, train_loss_step=4.56e+3, train_loss_epoch=4.67e+3]
Epoch 61/100:  60%|    | 60/100 [1:35:13&lt;1:03:13, 94.83s/it, v_num=1, train_loss_step=4.56e+3, train_loss_epoch=4.67e+3]
Epoch 61/100:  61%|    | 61/100 [1:36:48&lt;1:01:36, 94.78s/it, v_num=1, train_loss_step=4.56e+3, train_loss_epoch=4.67e+3]
Epoch 61/100:  61%|    | 61/100 [1:36:48&lt;1:01:36, 94.78s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 62/100:  61%|    | 61/100 [1:36:48&lt;1:01:36, 94.78s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 62/100:  62%|   | 62/100 [1:38:23&lt;1:00:02, 94.80s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 62/100:  62%|   | 62/100 [1:38:23&lt;1:00:02, 94.80s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.67e+3]
Epoch 63/100:  62%|   | 62/100 [1:38:23&lt;1:00:02, 94.80s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.67e+3]
Epoch 63/100:  63%|   | 63/100 [1:39:58&lt;58:29, 94.84s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.67e+3]  
Epoch 63/100:  63%|   | 63/100 [1:39:58&lt;58:29, 94.84s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.67e+3]
Epoch 64/100:  63%|   | 63/100 [1:39:58&lt;58:29, 94.84s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.67e+3]
Epoch 64/100:  64%|   | 64/100 [1:41:33&lt;56:56, 94.90s/it, v_num=1, train_loss_step=4.77e+3, train_loss_epoch=4.67e+3]
Epoch 64/100:  64%|   | 64/100 [1:41:33&lt;56:56, 94.90s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.67e+3] 
Epoch 65/100:  64%|   | 64/100 [1:41:33&lt;56:56, 94.90s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.67e+3]
Epoch 65/100:  65%|   | 65/100 [1:43:07&lt;55:17, 94.79s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.67e+3]
Epoch 65/100:  65%|   | 65/100 [1:43:07&lt;55:17, 94.79s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.67e+3]
Epoch 66/100:  65%|   | 65/100 [1:43:07&lt;55:17, 94.79s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.67e+3]
Epoch 66/100:  66%|   | 66/100 [1:44:43&lt;53:46, 94.91s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.67e+3]
Epoch 66/100:  66%|   | 66/100 [1:44:43&lt;53:46, 94.91s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 67/100:  66%|   | 66/100 [1:44:43&lt;53:46, 94.91s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 67/100:  67%|   | 67/100 [1:46:31&lt;54:27, 99.02s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 67/100:  67%|   | 67/100 [1:46:31&lt;54:27, 99.02s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 68/100:  67%|   | 67/100 [1:46:31&lt;54:27, 99.02s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 68/100:  68%|   | 68/100 [1:48:20&lt;54:23, 102.00s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.67e+3]
Epoch 68/100:  68%|   | 68/100 [1:48:20&lt;54:23, 102.00s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3]
Epoch 69/100:  68%|   | 68/100 [1:48:20&lt;54:23, 102.00s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3]
Epoch 69/100:  69%|   | 69/100 [1:49:51&lt;50:55, 98.57s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3] 
Epoch 69/100:  69%|   | 69/100 [1:49:51&lt;50:55, 98.57s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.66e+3]
Epoch 70/100:  69%|   | 69/100 [1:49:51&lt;50:55, 98.57s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.66e+3]
Epoch 70/100:  70%|   | 70/100 [1:51:21&lt;48:03, 96.11s/it, v_num=1, train_loss_step=4.64e+3, train_loss_epoch=4.66e+3]
Epoch 70/100:  70%|   | 70/100 [1:51:21&lt;48:03, 96.11s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 71/100:  70%|   | 70/100 [1:51:21&lt;48:03, 96.11s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 71/100:  71%|   | 71/100 [1:52:51&lt;45:35, 94.33s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 71/100:  71%|   | 71/100 [1:52:51&lt;45:35, 94.33s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=4.66e+3]
Epoch 72/100:  71%|   | 71/100 [1:52:51&lt;45:35, 94.33s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=4.66e+3]
Epoch 72/100:  72%|  | 72/100 [1:54:23&lt;43:35, 93.41s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=4.66e+3]
Epoch 72/100:  72%|  | 72/100 [1:54:23&lt;43:35, 93.41s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.66e+3]
Epoch 73/100:  72%|  | 72/100 [1:54:23&lt;43:35, 93.41s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.66e+3]
Epoch 73/100:  73%|  | 73/100 [1:55:57&lt;42:06, 93.58s/it, v_num=1, train_loss_step=4.69e+3, train_loss_epoch=4.66e+3]
Epoch 73/100:  73%|  | 73/100 [1:55:57&lt;42:06, 93.58s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 74/100:  73%|  | 73/100 [1:55:57&lt;42:06, 93.58s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 74/100:  74%|  | 74/100 [1:57:30&lt;40:35, 93.65s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 74/100:  74%|  | 74/100 [1:57:30&lt;40:35, 93.65s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.66e+3]
Epoch 75/100:  74%|  | 74/100 [1:57:30&lt;40:35, 93.65s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.66e+3]
Epoch 75/100:  75%|  | 75/100 [1:59:04&lt;39:03, 93.72s/it, v_num=1, train_loss_step=4.71e+3, train_loss_epoch=4.66e+3]
Epoch 75/100:  75%|  | 75/100 [1:59:04&lt;39:03, 93.72s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 76/100:  75%|  | 75/100 [1:59:04&lt;39:03, 93.72s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 76/100:  76%|  | 76/100 [2:00:39&lt;37:37, 94.05s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 76/100:  76%|  | 76/100 [2:00:39&lt;37:37, 94.05s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 77/100:  76%|  | 76/100 [2:00:39&lt;37:37, 94.05s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 77/100:  77%|  | 77/100 [2:02:16&lt;36:19, 94.78s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 77/100:  77%|  | 77/100 [2:02:16&lt;36:19, 94.78s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 78/100:  77%|  | 77/100 [2:02:16&lt;36:19, 94.78s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 78/100:  78%|  | 78/100 [2:04:06&lt;36:28, 99.48s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 78/100:  78%|  | 78/100 [2:04:06&lt;36:28, 99.48s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 79/100:  78%|  | 78/100 [2:04:06&lt;36:28, 99.48s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 79/100:  79%|  | 79/100 [2:05:58&lt;36:05, 103.13s/it, v_num=1, train_loss_step=4.66e+3, train_loss_epoch=4.66e+3]
Epoch 79/100:  79%|  | 79/100 [2:05:58&lt;36:05, 103.13s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3]
Epoch 80/100:  79%|  | 79/100 [2:05:58&lt;36:05, 103.13s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3]
Epoch 80/100:  80%|  | 80/100 [2:07:50&lt;35:15, 105.76s/it, v_num=1, train_loss_step=4.59e+3, train_loss_epoch=4.66e+3]
Epoch 80/100:  80%|  | 80/100 [2:07:50&lt;35:15, 105.76s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 81/100:  80%|  | 80/100 [2:07:50&lt;35:15, 105.76s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 81/100:  81%|  | 81/100 [2:09:42&lt;34:07, 107.75s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 81/100:  81%|  | 81/100 [2:09:42&lt;34:07, 107.75s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.66e+3]
Epoch 82/100:  81%|  | 81/100 [2:09:42&lt;34:07, 107.75s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.66e+3]
Epoch 82/100:  82%| | 82/100 [2:11:54&lt;34:32, 115.15s/it, v_num=1, train_loss_step=4.76e+3, train_loss_epoch=4.66e+3]
Epoch 82/100:  82%| | 82/100 [2:11:54&lt;34:32, 115.15s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3] 
Epoch 83/100:  82%| | 82/100 [2:11:54&lt;34:32, 115.15s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 83/100:  83%| | 83/100 [2:14:07&lt;34:09, 120.54s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 83/100:  83%| | 83/100 [2:14:07&lt;34:09, 120.54s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 84/100:  83%| | 83/100 [2:14:07&lt;34:09, 120.54s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 84/100:  84%| | 84/100 [2:16:19&lt;32:59, 123.72s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 84/100:  84%| | 84/100 [2:16:19&lt;32:59, 123.72s/it, v_num=1, train_loss_step=4.6e+3, train_loss_epoch=4.66e+3]
Epoch 85/100:  84%| | 84/100 [2:16:19&lt;32:59, 123.72s/it, v_num=1, train_loss_step=4.6e+3, train_loss_epoch=4.66e+3]
Epoch 85/100:  85%| | 85/100 [2:18:31&lt;31:36, 126.43s/it, v_num=1, train_loss_step=4.6e+3, train_loss_epoch=4.66e+3]
Epoch 85/100:  85%| | 85/100 [2:18:31&lt;31:36, 126.43s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 86/100:  85%| | 85/100 [2:18:31&lt;31:36, 126.43s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 86/100:  86%| | 86/100 [2:20:40&lt;29:38, 127.02s/it, v_num=1, train_loss_step=4.62e+3, train_loss_epoch=4.66e+3]
Epoch 86/100:  86%| | 86/100 [2:20:40&lt;29:38, 127.02s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 87/100:  86%| | 86/100 [2:20:40&lt;29:38, 127.02s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 87/100:  87%| | 87/100 [2:22:50&lt;27:44, 128.08s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 87/100:  87%| | 87/100 [2:22:50&lt;27:44, 128.08s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 88/100:  87%| | 87/100 [2:22:50&lt;27:44, 128.08s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 88/100:  88%| | 88/100 [2:25:00&lt;25:41, 128.47s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 88/100:  88%| | 88/100 [2:25:00&lt;25:41, 128.47s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 89/100:  88%| | 88/100 [2:25:00&lt;25:41, 128.47s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 89/100:  89%| | 89/100 [2:27:09&lt;23:36, 128.78s/it, v_num=1, train_loss_step=4.73e+3, train_loss_epoch=4.66e+3]
Epoch 89/100:  89%| | 89/100 [2:27:09&lt;23:36, 128.78s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 90/100:  89%| | 89/100 [2:27:09&lt;23:36, 128.78s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 90/100:  90%| | 90/100 [2:29:20&lt;21:34, 129.48s/it, v_num=1, train_loss_step=4.58e+3, train_loss_epoch=4.66e+3]
Epoch 90/100:  90%| | 90/100 [2:29:20&lt;21:34, 129.48s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3] 
Epoch 91/100:  90%| | 90/100 [2:29:20&lt;21:34, 129.48s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 91/100:  91%| | 91/100 [2:31:10&lt;18:30, 123.44s/it, v_num=1, train_loss_step=4.7e+3, train_loss_epoch=4.66e+3]
Epoch 91/100:  91%| | 91/100 [2:31:10&lt;18:30, 123.44s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 92/100:  91%| | 91/100 [2:31:10&lt;18:30, 123.44s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 92/100:  92%|| 92/100 [2:32:49&lt;15:28, 116.07s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 92/100:  92%|| 92/100 [2:32:49&lt;15:28, 116.07s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 93/100:  92%|| 92/100 [2:32:49&lt;15:28, 116.07s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 93/100:  93%|| 93/100 [2:34:28&lt;12:57, 111.05s/it, v_num=1, train_loss_step=4.63e+3, train_loss_epoch=4.66e+3]
Epoch 93/100:  93%|| 93/100 [2:34:28&lt;12:57, 111.05s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 94/100:  93%|| 93/100 [2:34:28&lt;12:57, 111.05s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 94/100:  94%|| 94/100 [2:36:08&lt;10:45, 107.64s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 94/100:  94%|| 94/100 [2:36:08&lt;10:45, 107.64s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 95/100:  94%|| 94/100 [2:36:08&lt;10:45, 107.64s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 95/100:  95%|| 95/100 [2:37:46&lt;08:44, 104.98s/it, v_num=1, train_loss_step=4.65e+3, train_loss_epoch=4.66e+3]
Epoch 95/100:  95%|| 95/100 [2:37:46&lt;08:44, 104.98s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.66e+3]
Epoch 96/100:  95%|| 95/100 [2:37:46&lt;08:44, 104.98s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.66e+3]
Epoch 96/100:  96%|| 96/100 [2:39:24&lt;06:51, 102.85s/it, v_num=1, train_loss_step=4.75e+3, train_loss_epoch=4.66e+3]
Epoch 96/100:  96%|| 96/100 [2:39:24&lt;06:51, 102.85s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 97/100:  96%|| 96/100 [2:39:24&lt;06:51, 102.85s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 97/100:  97%|| 97/100 [2:41:03&lt;05:05, 101.73s/it, v_num=1, train_loss_step=4.68e+3, train_loss_epoch=4.66e+3]
Epoch 97/100:  97%|| 97/100 [2:41:03&lt;05:05, 101.73s/it, v_num=1, train_loss_step=4.55e+3, train_loss_epoch=4.66e+3]
Epoch 98/100:  97%|| 97/100 [2:41:03&lt;05:05, 101.73s/it, v_num=1, train_loss_step=4.55e+3, train_loss_epoch=4.66e+3]
Epoch 98/100:  98%|| 98/100 [2:42:42&lt;03:21, 100.67s/it, v_num=1, train_loss_step=4.55e+3, train_loss_epoch=4.66e+3]
Epoch 98/100:  98%|| 98/100 [2:42:42&lt;03:21, 100.67s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.66e+3]
Epoch 99/100:  98%|| 98/100 [2:42:42&lt;03:21, 100.67s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.66e+3]
Epoch 99/100:  99%|| 99/100 [2:44:21&lt;01:40, 100.19s/it, v_num=1, train_loss_step=4.74e+3, train_loss_epoch=4.66e+3]
Epoch 99/100:  99%|| 99/100 [2:44:21&lt;01:40, 100.19s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 100/100:  99%|| 99/100 [2:44:21&lt;01:40, 100.19s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 100/100: 100%|| 100/100 [2:46:00&lt;00:00, 99.97s/it, v_num=1, train_loss_step=4.67e+3, train_loss_epoch=4.66e+3]
Epoch 100/100: 100%|| 100/100 [2:46:00&lt;00:00, 99.97s/it, v_num=1, train_loss_step=4.51e+3, train_loss_epoch=4.66e+3]
Epoch 100/100: 100%|| 100/100 [2:46:00&lt;00:00, 99.61s/it, v_num=1, train_loss_step=4.51e+3, train_loss_epoch=4.66e+3]
Removing group-batches with zero counts...

Number of cells per donor_id larger than 5000. Downsampling...
Preview:
An object of class Seurat 
22593 features across 19687 samples within 1 assay 
Active assay: RNA (22593 features, 2000 variable features)
 2 layers present: counts, data
 2 dimensional reductions calculated: scvi, umap

Counts by group and batch:
                                                            donor_id
cell_type                                                      D1   D2   D3
  endothelial cell of lymphatic vessel                         26   15   11
  basal-myoepithelial cell of mammary gland                   232   60  583
  luminal adaptive secretory precursor cell of mammary gland 1264  460 1109
  luminal hormone-sensing cell of mammary gland               777  329  809
                                                            donor_id
cell_type                                                      D4   D5  D11
  endothelial cell of lymphatic vessel                         10    4   19
  basal-myoepithelial cell of mammary gland                   174  326 2305
  luminal adaptive secretory precursor cell of mammary gland  842 1193 2560
  luminal hormone-sensing cell of mammary gland               744  719  116
                                                            donor_id
cell_type                                                    pooled [D9,D7,D8,D10,D6]
  endothelial cell of lymphatic vessel                                             15
  basal-myoepithelial cell of mammary gland                                       797
  luminal adaptive secretory precursor cell of mammary gland                     3941
  luminal hormone-sensing cell of mammary gland                                   247</code></pre>
</div>
</div>
<p>Note that scVI can also be run in supervised mode by providing cell type annotation. This is called scANVI.</p>
</section>
</section>
<section id="compare" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="compare"><span class="header-section-number">3</span> Compare</h2>
<section id="umap" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="umap"><span class="header-section-number">3.1</span> UMAP</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>path_umap_v <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path,<span class="fu">paste0</span>(label,<span class="st">"-umap-"</span>,grp,<span class="st">"-"</span>,suffix,<span class="st">".png"</span>))</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>ulist1 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(fn_plots, <span class="at">path =</span> path_rds, <span class="at">label =</span> method_names, <span class="at">path_umap =</span> path_umap_v, <span class="at">MoreArgs =</span> <span class="fu">list</span>(<span class="at">grp =</span> grp, <span class="at">cols =</span> cols_grp, <span class="at">export =</span> <span class="cn">TRUE</span>), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(<span class="at">plotlist =</span> ulist1, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">common.legend =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">annotate_figure</span>(<span class="at">p =</span> g, <span class="at">top =</span> <span class="fu">text_grob</span>(<span class="fu">paste0</span>(label,<span class="st">"  "</span>,grp), <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>, <span class="at">family =</span> <span class="st">"barlow"</span>, <span class="at">color =</span> <span class="st">"#2e4053"</span>))<span class="sc">+</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bgcolor</span>(<span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">border</span>(<span class="st">"white"</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-umap-"</span>, grp, <span class="st">".png"</span>)), g, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="fl">24.5</span>, <span class="at">width =</span> <span class="dv">24</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="breast/breast-integration-umap-cell_type.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="UMAP plots of various data integration methods. Cells colored by cell_type."><img src="breast/breast-integration-umap-cell_type.png" class="img-fluid figure-img" alt="UMAP plots of various data integration methods. Cells colored by cell_type."></a></p>
<figcaption>UMAP plots of various data integration methods. Cells colored by cell_type.</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>path_umap_v <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path,<span class="fu">paste0</span>(label,<span class="st">"-umap-"</span>,batch,<span class="st">"-"</span>,suffix,<span class="st">".png"</span>))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>ulist2 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(fn_plots, <span class="at">path=</span>path_rds, <span class="at">label=</span>method_names, <span class="at">path_umap =</span> path_umap_v, <span class="at">MoreArgs=</span><span class="fu">list</span>(<span class="at">grp=</span>batch, <span class="at">cols=</span>cols_batch, <span class="at">export=</span><span class="cn">TRUE</span>),<span class="at">SIMPLIFY=</span><span class="cn">FALSE</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(<span class="at">plotlist=</span>ulist2,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">nrow=</span><span class="dv">3</span>,<span class="at">common.legend=</span><span class="cn">TRUE</span>)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">annotate_figure</span>(<span class="at">p =</span> g, <span class="at">top =</span> <span class="fu">text_grob</span>(<span class="fu">paste0</span>(label,<span class="st">"  "</span>,batch), <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>, <span class="at">family =</span> <span class="st">"barlow"</span>, <span class="at">color =</span> <span class="st">"#2e4053"</span>))<span class="sc">+</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bgcolor</span>(<span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">border</span>(<span class="st">"white"</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path,<span class="fu">paste0</span>(label,<span class="st">"-integration-umap-"</span>,batch,<span class="st">".png"</span>)),g,<span class="at">dpi=</span><span class="dv">250</span>,<span class="at">height=</span><span class="dv">24</span>,<span class="at">width=</span><span class="dv">24</span>,<span class="at">units=</span><span class="st">"cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="breast/breast-integration-umap-donor_id.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="UMAP plots of various data integration methods. Cells colored by donor_id."><img src="breast/breast-integration-umap-donor_id.png" class="img-fluid figure-img" alt="UMAP plots of various data integration methods. Cells colored by donor_id."></a></p>
<figcaption>UMAP plots of various data integration methods. Cells colored by donor_id.</figcaption>
</figure>
</div>
</section>
<section id="metrics" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="metrics"><span class="header-section-number">3.2</span> Metrics</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>dfr <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(<span class="fu">mapply</span>(fn_metrics, <span class="at">path =</span> path_metrics, <span class="at">label =</span> method_names, <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>))</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> dfr <span class="sc">|&gt;</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">rev</span>(method_names))) <span class="sc">|&gt;</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">!</span>method, <span class="at">names_to =</span> <span class="st">"metric"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">|&gt;</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">metric =</span> <span class="fu">factor</span>(metric, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"iLISI"</span>,<span class="st">"norm_iLISI"</span>,<span class="st">"CiLISI"</span>,<span class="st">"CiLISI_means"</span>,<span class="st">"norm_cLISI"</span>,<span class="st">"norm_cLISI_means"</span>,<span class="st">"celltype_ASW"</span>,<span class="st">"celltype_ASW_means"</span>))) <span class="sc">|&gt;</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(method, value)) <span class="sc">+</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>metric, <span class="at">scales =</span> <span class="st">"free_x"</span>, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-metrics.png"</span>)), g, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">15</span>, <span class="at">width =</span> <span class="dv">24</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> dfr <span class="sc">|&gt;</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">rev</span>(method_names))) <span class="sc">|&gt;</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">!</span>method, <span class="at">names_to =</span> <span class="st">"metric"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">|&gt;</span></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(metric <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"norm_iLISI"</span>,<span class="st">"norm_cLISI"</span>,<span class="st">"celltype_ASW"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">metric =</span> <span class="fu">factor</span>(metric, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"norm_iLISI"</span>,<span class="st">"norm_cLISI"</span>,<span class="st">"celltype_ASW"</span>))) <span class="sc">|&gt;</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(method, value)) <span class="sc">+</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>metric, <span class="at">scales =</span> <span class="st">"free_x"</span>, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-metrics-slim.png"</span>)), g, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">20</span>, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="breast/breast-integration-metrics.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Various integration metrics across methods."><img src="breast/breast-integration-metrics.png" class="img-fluid figure-img" alt="Various integration metrics across methods."></a></p>
<figcaption>Various integration metrics across methods.</figcaption>
</figure>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> dfr <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  GGally<span class="sc">::</span><span class="fu">ggpairs</span>(<span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>,<span class="at">upper =</span> <span class="fu">list</span>(<span class="at">continuous =</span> <span class="fu">wrap</span>(<span class="st">"cor"</span>, <span class="at">size =</span> <span class="dv">2</span>)),<span class="at">lower=</span><span class="fu">list</span>(<span class="at">continuous =</span> <span class="fu">wrap</span>(<span class="st">"points"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">size=</span><span class="fl">0.6</span>)))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p<span class="sc">+</span><span class="fu">theme_report</span>()</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-metrics-pairs.png"</span>)), p, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">20</span>, <span class="at">width =</span> <span class="dv">20</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="breast/breast-integration-metrics-pairs.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Correlation between integration metrics."><img src="breast/breast-integration-metrics-pairs.png" class="img-fluid figure-img" alt="Correlation between integration metrics."></a></p>
<figcaption>Correlation between integration metrics.</figcaption>
</figure>
</div>
</section>
<section id="time-memory" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="time-memory"><span class="header-section-number">3.3</span> Time &amp; Memory</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>dfr <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(<span class="fu">mapply</span>(fn_memtime, <span class="at">path =</span> path_memtime, <span class="at">label =</span> method_names, <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)) <span class="sc">|&gt;</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(time_sec) <span class="sc">|&gt;</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time_min =</span> <span class="fu">round</span>(time_sec <span class="sc">/</span> <span class="dv">60</span>, <span class="dv">2</span>))</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> dfr <span class="sc">%&gt;%</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(time_min) <span class="sc">|&gt;</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">unique</span>(method))) <span class="sc">|&gt;</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(method, time_min)) <span class="sc">+</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> <span class="st">"log1p"</span>, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">30</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">150</span>, <span class="dv">200</span>)) <span class="sc">+</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"Log Time (Min)"</span>) <span class="sc">+</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-time.png"</span>)), g1, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">8</span>, <span class="at">width =</span> <span class="dv">14</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> dfr <span class="sc">%&gt;%</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(peak_ram_mb) <span class="sc">|&gt;</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">unique</span>(method))) <span class="sc">|&gt;</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(method, peak_ram_mb)) <span class="sc">+</span></span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># scale_y_continuous(trans="log1p",breaks=c(1,2,3,4,5,10,15,30,50,100,150))+</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"Peak RAM used (MB)"</span>) <span class="sc">+</span></span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-mem-peak.png"</span>)), g2, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">8</span>, <span class="at">width =</span> <span class="dv">14</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a>g3 <span class="ot">&lt;-</span> dfr <span class="sc">%&gt;%</span></span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(total_ram_mb) <span class="sc">|&gt;</span></span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">unique</span>(method))) <span class="sc">|&gt;</span></span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(method, total_ram_mb)) <span class="sc">+</span></span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># scale_y_continuous(trans="log1p",breaks=c(1,2,3,4,5,10,15,30,50,100,150))+</span></span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"Total RAM used (MB)"</span>) <span class="sc">+</span></span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-mem-total.png"</span>)), g3, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">8</span>, <span class="at">width =</span> <span class="dv">14</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a>g4 <span class="ot">&lt;-</span> dfr <span class="sc">%&gt;%</span></span>
<span id="cb36-40"><a href="#cb36-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> method_names)) <span class="sc">|&gt;</span></span>
<span id="cb36-41"><a href="#cb36-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(time_min, peak_ram_mb)) <span class="sc">+</span></span>
<span id="cb36-42"><a href="#cb36-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb36-43"><a href="#cb36-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> method), <span class="at">size =</span> <span class="dv">3</span>, <span class="at">nudge_x =</span> <span class="fl">0.2</span>, <span class="at">hjust =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb36-44"><a href="#cb36-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">trans =</span> <span class="st">"log1p"</span>, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">30</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">150</span>, <span class="dv">200</span>)) <span class="sc">+</span></span>
<span id="cb36-45"><a href="#cb36-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Peak RAM used (MB)"</span>, <span class="at">x =</span> <span class="st">"Log Time (Min)"</span>) <span class="sc">+</span></span>
<span id="cb36-46"><a href="#cb36-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_report</span>()</span>
<span id="cb36-47"><a href="#cb36-47" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-memtime-scatter.png"</span>)), g4, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">15</span>, <span class="at">width =</span> <span class="dv">15</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span>
<span id="cb36-48"><a href="#cb36-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-49"><a href="#cb36-49" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> ggpubr<span class="sc">::</span><span class="fu">ggarrange</span>(g1, g2, g3, g4, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb36-50"><a href="#cb36-50" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">file.path</span>(path, <span class="fu">paste0</span>(label, <span class="st">"-integration-memtime.png"</span>)), g, <span class="at">dpi =</span> <span class="dv">250</span>, <span class="at">height =</span> <span class="dv">20</span>, <span class="at">width =</span> <span class="dv">22</span>, <span class="at">units =</span> <span class="st">"cm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="breast/breast-integration-memtime.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="(Top left) Total time taken, (Top right) peak memory usage, (Bottom left) and total memory usage for each method. (Bottom right) Scatterplot between peak memory usage and time."><img src="breast/breast-integration-memtime.png" class="img-fluid figure-img" alt="(Top left) Total time taken, (Top right) peak memory usage, (Bottom left) and total memory usage for each method. (Bottom right) Scatterplot between peak memory usage and time."></a></p>
<figcaption>(Top left) Total time taken, (Top right) peak memory usage, (Bottom left) and total memory usage for each method. (Bottom right) Scatterplot between peak memory usage and time.</figcaption>
</figure>
</div>
</section>
</section>
<section id="versions" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="versions"><span class="header-section-number">4</span> Versions</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"Seurat: "</span>, <span class="fu">packageVersion</span>(<span class="st">"Seurat"</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"harmony: "</span>, <span class="fu">packageVersion</span>(<span class="st">"harmony"</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"rliger: "</span>, <span class="fu">packageVersion</span>(<span class="st">"rliger"</span>))</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"batchelor: "</span>, <span class="fu">packageVersion</span>(<span class="st">"batchelor"</span>))</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"STACAS: "</span>, <span class="fu">packageVersion</span>(<span class="st">"STACAS"</span>))</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"conos: "</span>, <span class="fu">packageVersion</span>(<span class="st">"conos"</span>))</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Sys.setenv(PYTHONPATH="/home/rstudio/.local/lib/python3.10/site-packages")</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">PYTHONPATH=</span><span class="st">"/usr/lib/python3.10/site-packages"</span>)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>scanorama <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scanorama"</span>)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>scvi <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"scvi"</span>, <span class="at">convert =</span> <span class="cn">FALSE</span>)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"scanorama: "</span>, scanorama<span class="sc">$</span><span class="st">`</span><span class="at">__version__</span><span class="st">`</span>)</span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(<span class="st">"scvi-tools: "</span>, scvi<span class="sc">$</span><span class="st">`</span><span class="at">__version__</span><span class="st">`</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Seurat: 5.1.0"
[1] "harmony: 1.2.0"
[1] "rliger: 2.0.1"
[1] "batchelor: 1.20.0"
[1] "STACAS: 2.2.2"
[1] "conos: 1.5.2"
[1] "scanorama: 1.7.4"
[1] "scvi-tools: 1.1.6"</code></pre>
</div>
</div>
</section>
<section id="session" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="session"><span class="header-section-number">5</span> Session</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] GGally_2.2.1                dplyr_1.1.4                
 [3] ggpubr_0.6.0                sceasy_0.0.7               
 [5] reticulate_1.39.0           conos_1.5.2                
 [7] igraph_2.0.3                Matrix_1.7-0               
 [9] STACAS_2.2.2                batchelor_1.20.0           
[11] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
[13] Biobase_2.64.0              GenomicRanges_1.56.1       
[15] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[17] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[19] MatrixGenerics_1.16.0       matrixStats_1.4.1          
[21] rliger_2.0.1                harmony_1.2.0              
[23] Rcpp_1.0.13                 showtext_0.9-7             
[25] showtextdb_3.0              sysfonts_0.8.9             
[27] plummy_0.2                  scIntegrationMetrics_1.1   
[29] SeuratWrappers_0.3.5        Seurat_5.1.0               
[31] SeuratObject_5.0.2          sp_2.1-4                   
[33] peakRAM_1.0.3               randomcoloR_1.1.0.1        
[35] glue_1.7.0                  ggplot2_3.5.1              
[37] stringr_1.5.1              

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0     httr_1.4.7               
  [3] RColorBrewer_1.1-3        doParallel_1.0.17        
  [5] backports_1.5.0           tools_4.4.1              
  [7] sctransform_0.4.1         utf8_1.2.4               
  [9] R6_2.5.1                  vegan_2.6-8              
 [11] ResidualMatrix_1.14.1     lazyeval_0.2.2           
 [13] uwot_0.2.2                mgcv_1.9-1               
 [15] GetoptLong_1.0.5          permute_0.9-7            
 [17] withr_3.0.1               gridExtra_2.3            
 [19] progressr_0.14.0          cli_3.6.3                
 [21] textshaping_0.4.0         spatstat.explore_3.3-2   
 [23] fastDummies_1.7.4         labeling_0.4.3           
 [25] spatstat.data_3.1-2       ggridges_0.5.6           
 [27] pbapply_1.7-2             systemfonts_1.1.0        
 [29] R.utils_2.12.3            parallelly_1.38.0        
 [31] generics_0.1.3            shape_1.4.6.1            
 [33] ica_1.0-3                 spatstat.random_3.3-1    
 [35] car_3.1-2                 fansi_1.0.6              
 [37] abind_1.4-8               R.methodsS3_1.8.2        
 [39] lifecycle_1.0.4           yaml_2.3.10              
 [41] carData_3.0-5             SparseArray_1.4.8        
 [43] Rtsne_0.17                glmGamPoi_1.16.0         
 [45] grid_4.4.1                promises_1.3.0           
 [47] crayon_1.5.3              miniUI_0.1.1.1           
 [49] lattice_0.22-6            beachmat_2.20.0          
 [51] cowplot_1.1.3             pillar_1.9.0             
 [53] knitr_1.48                ComplexHeatmap_2.20.0    
 [55] rjson_0.2.23              future.apply_1.11.2      
 [57] codetools_0.2-20          leiden_0.4.3.1           
 [59] V8_5.0.0                  spatstat.univar_3.0-1    
 [61] data.table_1.16.0         remotes_2.5.0            
 [63] RcppPlanc_1.0.0           vctrs_0.6.5              
 [65] png_0.1-8                 spam_2.10-0              
 [67] gtable_0.3.5              xfun_0.47                
 [69] S4Arrays_1.4.1            mime_0.12                
 [71] survival_3.7-0            iterators_1.0.14         
 [73] fitdistrplus_1.2-1        ROCR_1.0-11              
 [75] nlme_3.1-166              RcppAnnoy_0.0.22         
 [77] rprojroot_2.0.4           irlba_2.3.5.1            
 [79] KernSmooth_2.23-24        colorspace_2.1-1         
 [81] tidyselect_1.2.1          compiler_4.4.1           
 [83] curl_5.2.2                BiocNeighbors_1.22.0     
 [85] DelayedArray_0.30.1       plotly_4.10.4            
 [87] scales_1.3.0              lmtest_0.9-40            
 [89] rappdirs_0.3.3            digest_0.6.37            
 [91] goftest_1.2-3             spatstat.utils_3.1-0     
 [93] rmarkdown_2.28            XVector_0.44.0           
 [95] RhpcBLASctl_0.23-42       htmltools_0.5.8.1        
 [97] pkgconfig_2.0.3           sparseMatrixStats_1.16.0 
 [99] fastmap_1.2.0             rlang_1.1.4              
[101] GlobalOptions_0.1.2       htmlwidgets_1.6.4        
[103] UCSC.utils_1.0.0          shiny_1.9.1              
[105] DelayedMatrixStats_1.26.0 farver_2.1.2             
[107] zoo_1.8-12                jsonlite_1.8.8           
[109] BiocParallel_1.38.0       R.oo_1.26.0              
[111] BiocSingular_1.20.0       magrittr_2.0.3           
[113] scuttle_1.14.0            GenomeInfoDbData_1.2.12  
[115] dotCall64_1.1-1           patchwork_1.2.0          
[117] munsell_0.5.1             leidenAlg_1.1.3          
[119] stringi_1.8.4             zlibbioc_1.50.0          
[121] MASS_7.3-61               plyr_1.8.9               
[123] ggstats_0.6.0             parallel_4.4.1           
[125] listenv_0.9.1             ggrepel_0.9.6            
[127] deldir_2.0-4              sccore_1.0.5             
[129] splines_4.4.1             tensor_1.5               
[131] circlize_0.4.16           spatstat.geom_3.3-2      
[133] ggsignif_0.6.4            RcppHNSW_0.6.0           
[135] reshape2_1.4.4            ScaledMatrix_1.12.0      
[137] evaluate_0.24.0           BiocManager_1.30.25      
[139] foreach_1.5.2             httpuv_1.6.15            
[141] RANN_2.6.2                tidyr_1.3.1              
[143] purrr_1.0.2               polyclip_1.10-7          
[145] future_1.34.0             clue_0.3-65              
[147] scattermore_1.2           rsvd_1.0.5               
[149] broom_1.0.6               xtable_1.8-4             
[151] RSpectra_0.16-2           rstatix_0.7.2            
[153] later_1.3.2               viridisLite_0.4.2        
[155] ragg_1.3.3                tibble_3.2.1             
[157] cluster_2.1.6             globals_0.16.3           
[159] here_1.0.1               </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","openEffect":"zoom","selector":".lightbox","loop":false,"descPosition":"bottom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>